### DISCUSSION TITLE
C# Assembly content type for mods

### DISCUSSION THREAD

**Discussion by Weilbyte** [FakeFishGames/Barotrauma#6393] [Ideas and suggestions]
Since modding appears to be limited to XML, it would be great if there was an ability to load a C# assembly, so modders can add things in like custom components, and maybe even edit core parts of the game with tools like Harmony.

There could be a new ContentType called `Assembly`, which when a mod gets loaded, the Assembly file gets loaded into the AppDomain.

Maybe something like this?
```cs
public void EnableContentPackageItems(IEnumerable<ContentFile> unorderedFiles)
        {
            if (WaitingForAutoUpdate) { return; }
            IOrderedEnumerable<ContentFile> files = unorderedFiles.OrderBy(ContentFileLoadOrder);
            foreach (ContentFile file in files)
            {
                switch (file.Type)
                {
                    case ContentType.Assembly:
                        DebugConsole.NewMessage("Hit Assembly content type " + file.Path);
                        Assembly asm;
                        try
                        {
                            byte[] rawAssembly = File.ReadAllBytes(file.Path);
                            asm = AppDomain.CurrentDomain.Load(rawAssembly);
                        }
                        catch (Exception ex)
                        {
                            DebugConsole.NewMessage("Exception loading " + file.Path + ": " + ex.ToString());
                        }
                        break;
                    case ContentType.Character:
                        ...
                        ...
```

And then for example, in the item's XML file a modder could include their component as `ModExample.Components.Example`. Something like this would shoddily detect that its specifying assembly name and load it perhaps?

```cs
 public static ItemComponent Load(XElement element, Item item, string file, bool errorMessages = true)
        {
            Type t;
            string type = element.Name.ToString().ToLowerInvariant();
            try
            {
                // Get the type of a specified class.
                if (type.Contains("."))
                {
                    t = Type.GetType(type, false, true); // If name contains a dot, assume name is assembly type.
                } else {
                    t = Type.GetType("Barotrauma.Items.Components." + type + "", false, true);
                }

                ...
                ...
```

But a bigger issue is that C# mods cant use important parts of the Barotrauma dll because they are access restricted, I do not have any clue about how this can be "fixed".

I don't have much experience with reflection and AppDomain but maybe something like this could work? Maybe just a shot in the dark, but would be great if this game comes out to be as moddable as RimWorld. Thoughts?

**Comment by mc-oofert** [comment]
watch me mod in the memz virus into a barotrauma mod lmfao

**Reply by Weilbyte** [reply]
that would not be very poggers
